[Caffe2] Fix bug in `str` on wide types

JoostHouben opened the pull request:
Summary:
The current implementation of `str` passes wide types (`wchar_t`, `wchar_t*`, `std::wstring`) directly to `std::ostringstream`. This has the following behavior:

 - C++17, `wchar_t` & `wchar_t *`: print the integer representation of the character or the pointer. This is unexpected and almost certainly a (runtime) bug.
 - C++17, `std::wstring`: compile-time error.
 - C++20, all of the above: compile-time error.

To fix the bug and to enable C++20 migration, this diff performs narrowing on these wide types (assuming UTF-16 encoding) before passing them to `std::ostringstream`. This fixes both the C++20 compile time errors and the C++17 runtime bugs.

This bug surfaced in enabling C++20 windows builds, because windows specific caffe2 code uses `TORCH_CHECK` with wide strings, which references `str` for generating error messages.

Test Plan: CI & https://godbolt.org/z/ecTGd8Ma9

Differential Revision: D52792393

malfet commented on c10/util/StringUtil.cpp:
```diff
@@ -1,7 +1,16 @@
 #include <string>
 
+#ifdef _WIN32
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
```
```suggestion
#include <c10/util/win32-headers.h>
```

malfet commented on c10/util/StringUtil.cpp:
```diff
@@ -28,6 +37,94 @@ std::string ExcludeFileExtension(const std::string& file_name) {
+// The WIN32 implementation of wstring_convert leaks memory; see
+// https://github.com/microsoft/STL/issues/443
+
+template <>
```
This feels like a huge code duplication from https://github.com/pytorch/pytorch/blob/208e64a9ba2abca63ea77b3b2dc5a2f19e4c2ac4/c10/util/Unicode.cpp#L26-L34 can perhaps template from Unicode.cpp be extended and used here?

JoostHouben committed:
[Caffe2] Fix bug in `str` on wide types (#117531)

Summary:

The current implementation of `str` passes wide types (`wchar_t`, `wchar_t*`, `std::wstring`) directly to `std::ostringstream`. This has the following behavior:

 - C++17, `wchar_t` & `wchar_t *`: print the integer representation of the character or the pointer. This is unexpected and almost certainly a (runtime) bug.
 - C++17, `std::wstring`: compile-time error.
 - C++20, all of the above: compile-time error.

To fix the bug and to enable C++20 migration, this diff performs narrowing on these wide types (assuming UTF-16 encoding) before passing them to `std::ostringstream`. This fixes both the C++20 compile time errors and the C++17 runtime bugs.

This bug surfaced in enabling C++20 windows builds, because windows specific caffe2 code uses `TORCH_CHECK` with wide strings, which references `str` for generating error messages.

Test Plan: CI

Reviewed By: johnkearney

Differential Revision: D52792393
